Modules/_zstd/compressor.c

```diff
@@ -153,8 +157,6 @@ _get_CDict(ZstdDict *self, int compressionLevel)
     PyObject *capsule;
     ZSTD_CDict *cdict;
 
-    // TODO(emmatyping): refactor critical section code into a lock_held function
```

Comment:
I think `_get_CDict` needs some more work. It doesn't look thread-safe to me (even with the GIL) because the `Py_BEGIN_ALLOW_THREADS` below means multiple threads can try to create the capsule for the same level.

I can think of two high-level strategies:

1) Use a lock on `ZstdDict` to ensure that only one thread tries to initialize the `cdict` at a time.

2) Use `PyDict_SetDefaultRef` so that only one thread succeeds in setting the capsule on the dict.

Is `self->dict_content` mutable? If so, then the lock may be more appropriate.

Reply:
Yeah that's a great point! I think `PyDict_SetDefaultRef` sounds more reasonable, the content is immutable so it should be safe to use `PyDict_SetDefaultRef`.

Reply:
I've switched to using PyDict_SetDefaultRef and PyDict_GetItemRef. I also added a test that should check sharing a dictionary across threads with several levels.

Reply:
Okay I actually used locking in the end, because we also want to make sure the member access of `ZstdDict->d_dict` is thread safe.